grade-management-new/GradeManagement.Services/Services/SubjectService.cs

```diff
@@ -0,0 +1,80 @@
+using AutoMapper;
+using GradeManagement.Data.Data;
+using GradeManagement.Data.Models;
+using GradeManagement.Shared.DTOs;
+using Microsoft.EntityFrameworkCore;
+using Task = System.Threading.Tasks.Task;
+
+namespace GradeManagement.Services.Services;
+
+public class SubjectService
+{
+    private readonly GradeManagementDbContext _gradeManagementDbContext;
+    private readonly IMapper _mapper;
+
+    public SubjectService(GradeManagementDbContext gradeManagementDbContext, IMapper mapper)
+    {
+        _gradeManagementDbContext = gradeManagementDbContext;
+        _mapper = mapper;
+    }
+
+    public async Task<IEnumerable<SubjectDTO>> GetAllSubjects()
+    {
+        List<Subject> allSubjects = await _gradeManagementDbContext.Subject.OrderByDescending(s => s.Id).ToListAsync();
+        var subjectDtos = _mapper.Map<List<SubjectDTO>>(allSubjects);
+        return subjectDtos;
+    }
+
+    public async Task<SubjectDTO> GetSubjectById(long id)
+    {
+        Subject? subject = await _gradeManagementDbContext.Subject.SingleOrDefaultAsync(s => s.Id == id);
+        if (subject == null) throw new NullReferenceException("No subject could be found with the given Id!");
+        var subjectDto = _mapper.Map<SubjectDTO>(subject);
+        return subjectDto;
+    }
+
+    public async Task<SubjectDTO> UpdateSubject(long id, SubjectDTO subjectDto)
+    {
+        if (subjectDto.Id != id) throw new ArgumentException("The Id from the query and the Id of the DTO do not match!");
+        Subject? subject = await _gradeManagementDbContext.Subject.Include(s=>s.Courses).SingleOrDefaultAsync(s => s.Id == id);
+        if (subject == null) throw new NullReferenceException("No subject could be found with the given Id!");
+        subject.Name = subjectDto.Name;
+        subject.NeptunSubjectCode = subjectDto.NeptunSubjectCode;
+
+        var courses = _gradeManagementDbContext.Course
+            .Where(c => subjectDto.CourseDtos.Select(co => co.Id).Contains(c.Id)).ToList();
+        List<Course> coursesToDelete = new List<Course>();
+        foreach (var course in subject.Courses)
```

Comment:
Itt nekem nem tiszta mi lesz a UI logika, hogy ezt az esetet is le kelljen kezelni

Reply:
Refaktorálásra vár a UI igényei alapján